lib/ts/BufferWriter.h

```diff
@@ -0,0 +1,326 @@
+  //
+  virtual ~BufferWriter() {}
+};
+
+// A buffer writer that writes to an array of char that is external to the writer instance.
+//
+class FixedBufferWriter : public BufferWriter
+{
+protected:
+  FixedBufferWriter(char *buf, size_t capacity, size_t attempted) : _buf(buf), _capacity(capacity), _attempted(attempted) {}
+
+public:
+  // 'buf' is a pointer to the external array of char to write to.  'capacity' is the number of bytes in the array.
+  //
+  // If you create a instance of this class with capacity == 0 (and a nullptr buffer), you can use it to measure the number of
+  // characters a series of writes would result it (from the extent() value) without actually writing.
+  //
+  FixedBufferWriter(char *buf, size_t capacity) : FixedBufferWriter(buf, capacity, 0) {}
+
+  FixedBufferWriter &
+  write(char c) override
+  {
+    if (_attempted < _capacity) {
+      _buf[_attempted] = c;
+    }
+    ++_attempted;
+
+    return *this;
+  }
+
+  FixedBufferWriter &
+  write(const void *data, size_t length) override
+  {
+    size_t newSize = _attempted + length;
+
+    if (newSize <= _capacity) {
+      std::memcpy(_buf + _attempted, data, length);
+
+    } else if (_attempted < _capacity) {
+      std::memcpy(_buf + _attempted, data, _capacity - _attempted);
+    }
+    _attempted = newSize;
+
+    return *this;
+  }
+
+  // It's not clear to my why g++ needs this using declaration in order to consider the inherited versions of 'write' when
+  // resolving calls to a 'write' member ( wkaras@oath.com ).
+  //
+  using BufferWriter::write;
+
+  bool
+  error() const override
+  {
+    return _attempted > _capacity;
+  }
+
+  char *
+  auxBuffer() override
+  {
+    return error() ? nullptr : _buf + _attempted;
+  }
+
+  FixedBufferWriter &
+  write(size_t n) override
+  {
+    _attempted += n;
+
+    return *this;
+  }
+
+  size_t
+  capacity() const override
+  {
+    return _capacity;
+  }
+
+  size_t
+  extent() const override
+  {
+    return _attempted;
+  }
+
+  FixedBufferWriter &
+  clip(size_t n) override
+  {
+    ink_assert(n <= _capacity);
+
+    _capacity -= n;
+
+    return *this;
+  }
+
+  FixedBufferWriter &
+  extend(size_t n) override
+  {
+    if (error()) {
+      _attempted = _capacity;
+    }
+
+    _capacity += n;
+
+    return *this;
+  }
+
+  // Reduce extent.  If extent is less than capacity, error condition is cleared.
+  //
+  void
+  reduce(size_t smallerExtent)
+  {
+    ink_assert(smallerExtent <= _attempted);
+
+    _attempted = smallerExtent;
+  }
+
+  // Provide a string_view of all successfully written characters.
+  //
+  string_view
+  view() const
+  {
+    return string_view(_buf, size());
+  }
+
+  operator string_view() const { return view(); }
+
+  // No copying
+  //
+  FixedBufferWriter(const FixedBufferWriter &) = delete;
+  FixedBufferWriter &operator=(const FixedBufferWriter &) = delete;
+
+  // Moving is OK.
+  //
+  FixedBufferWriter(FixedBufferWriter &&) = default;
```

Comment:
The main purpose of move constructors is when the source of the move is a nameless temporary implicitly created by the compiler. After the move, it will not be used again before its destructor is called.  The move must alter the source in order to make sure its destructor call does not corrupt the destination.  Typically that means the destructor must not destroy/release resources (such as memory) whose ownership has been transferred to the destination of the move.  When named objects are the source of a move (using std::move), the code doing this should make sure the object is destroyed as the next operation on it, or some other operation is done on the object, to make it safe to continue to use.